src/core/component.js

```diff
@@ -111,63 +112,27 @@ Component.prototype = {
    * @returns {object} Component data.
    */
   parse: function (value) {
-    var typeName;
-    var type;
     var schema = this.schema;
-    var properties;
-
-    if (isSingleProp(this.schema)) {
-      typeName = this.schema.type;
-      type = propertyTypes[typeName];
-      if (type) { return type.parse.call(this, value); }
-      return error(typeName + ' is not a valid type.');
-    }
-
-    properties = objectParse(value);
-
-    if (properties === null || typeof properties !== 'object') { return properties; }
-
-    Object.keys(properties).forEach(parseProperty);
-    return properties;
-
-    function parseProperty (key) {
-      if (!schema[key]) { return; }
-      properties[key] = schema[key].parse(properties[key]);
-    }
+    if (isSingleProp(schema)) { return parseProperty(value, schema); }
+    return parseProperties(objectParse(value), schema, true);
   },
 
   /**
-   * Stringifies each property based on property type.
-   * If component is single-property, then stringifies the single property value.
+   * Stringify properties if necessary.
    *
-   * @param {object} data
+   * Only called from `entity.setAttribute` for properties that accept an object value such as
+   * vec3 {x, y, z}.
+   *
+   * @param {object} data - Complete component data.
    * @returns {string}
    */
   stringify: function (data) {
-    var typeName;
-    var type;
-    var processedData;
     var schema = this.schema;
+    if (typeof data === 'string') { return data; }
 
-    if (isSingleProp(schema)) {
-      typeName = schema.type;
-      type = propertyTypes[typeName];
-      if (type) { return type.stringify.call(this, data); }
-      return error(typeName + ' is not a valid type.');
-    }
-
-    if (typeof data !== 'object') { return data; }
-
-    processedData = utils.extend({}, data);
-
-    Object.keys(processedData).forEach(stringifyProperty);
-
-    return objectStringify(processedData);
-
-    function stringifyProperty (key) {
-      if (!schema[key]) { return; }
-      processedData[key] = schema[key].stringify(processedData[key]);
-    }
+    if (isSingleProp(schema)) { return stringifyProperty(data, schema); }
```

Comment:
that already happens at https://github.com/aframevr/aframe/blob/dev/src/core/schema.js#L63